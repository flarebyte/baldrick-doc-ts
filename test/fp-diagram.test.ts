import {
  mergeSourceDiagrams,
  toFunctionalProgrammingDiagram,
  toFunctionalProgrammingMermaid,
} from '../src/fp-diagram';
import {
  FunctionDiagram,
  FunctionInfo,
  ImportInfo,
  ModuleInfo,
  SourceDiagram,
  SourceInfo,
} from '../src/parser-model';

const createImportInfo = (identifier: string, from: string): ImportInfo => ({
  identifier,
  from,
});

const createFunctionInfo = (
  identifier: string,
  exported: boolean
): FunctionInfo => ({
  identifier,
  bodyWidth: identifier.length,
  keywords: ['keyword1', `keyword-${identifier}`],
  description: `Description of ${identifier}`,
  exported,
  descendantCount: 100,
  expression: exported,
});

const createSource = (id: number): SourceInfo => ({
  filename: `filename-${id}.js`,
  imports: [createImportInfo(`import-${id}-function1`, `external-${id}`)],
  functions: [
    createFunctionInfo(`create-${id}`, true),
    createFunctionInfo(`update-${id}`, false),
    createFunctionInfo(`delete-${id}`, true),
  ],
});

const moduleInfo: ModuleInfo = {
  name: 'this-project',
  sources: [
    createSource(7),
    createSource(11),
    createSource(17),
    { ...createSource(21), functions: [] },
  ],
};

const createFunctionDiagram = (identifier: string): FunctionDiagram => ({
  identifier,
  exported: identifier.startsWith('exp'),
});

const createSourceDiagram = (
  filename: string,
  functions: string[]
): SourceDiagram => ({
  filename,
  external: true,
  functions: functions.map(createFunctionDiagram),
});

describe('fp-diagram', () => {
  it('should convert diagram to a diagram structure', () => {
    const actual = toFunctionalProgrammingDiagram(moduleInfo);
    expect(actual).toMatchSnapshot();
    const mermaid = toFunctionalProgrammingMermaid(actual);
    expect(mermaid).toMatchSnapshot();
  });
  it('merge source diagrams', () => {
    const sources = [
      createSourceDiagram('Earth', ['Moon']),
      createSourceDiagram('Earth', ['Moon']),
      createSourceDiagram('Jupiter', ['Europa']),
      createSourceDiagram('Mercury', []),
      createSourceDiagram('Jupiter', ['Ganymede', 'Europa']),
      createSourceDiagram('Jupiter', ['Callisto']),
      createSourceDiagram('Jupiter', ['Callisto', 'Himalia']),
      createSourceDiagram('Jupiter', ['Callisto', 'Himalia']),
      createSourceDiagram('Jupiter', ['Elara']),
    ];
    const actual = mergeSourceDiagrams(sources);
    expect(actual).toMatchInlineSnapshot(`
      Array [
        Object {
          "external": true,
          "filename": "Earth",
          "functions": Array [
            Object {
              "exported": false,
              "identifier": "Moon",
            },
          ],
        },
        Object {
          "external": true,
          "filename": "Jupiter",
          "functions": Array [
            Object {
              "exported": false,
              "identifier": "Elara",
            },
            Object {
              "exported": false,
              "identifier": "Callisto",
            },
            Object {
              "exported": false,
              "identifier": "Himalia",
            },
            Object {
              "exported": false,
              "identifier": "Ganymede",
            },
            Object {
              "exported": false,
              "identifier": "Europa",
            },
          ],
        },
        Object {
          "external": true,
          "filename": "Mercury",
          "functions": Array [],
        },
      ]
    `);
  });
});
