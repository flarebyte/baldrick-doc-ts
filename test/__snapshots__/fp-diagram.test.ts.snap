// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fp-diagram should convert diagram to a diagram structure 1`] = `
Object {
  "entities": Array [
    Object {
      "external": false,
      "filename": "filename-7.js",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "create-7",
        },
        Object {
          "exported": false,
          "identifier": "update-7",
        },
        Object {
          "exported": true,
          "identifier": "delete-7",
        },
      ],
    },
    Object {
      "external": false,
      "filename": "filename-11.js",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "create-11",
        },
        Object {
          "exported": false,
          "identifier": "update-11",
        },
        Object {
          "exported": true,
          "identifier": "delete-11",
        },
      ],
    },
    Object {
      "external": false,
      "filename": "filename-17.js",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "create-17",
        },
        Object {
          "exported": false,
          "identifier": "update-17",
        },
        Object {
          "exported": true,
          "identifier": "delete-17",
        },
      ],
    },
    Object {
      "external": false,
      "filename": "filename-21.js",
      "functions": Array [],
    },
    Object {
      "external": true,
      "filename": "external-7",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "import-7-function1",
        },
      ],
    },
    Object {
      "external": true,
      "filename": "external-11",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "import-11-function1",
        },
      ],
    },
    Object {
      "external": true,
      "filename": "external-17",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "import-17-function1",
        },
      ],
    },
    Object {
      "external": true,
      "filename": "external-21",
      "functions": Array [
        Object {
          "exported": true,
          "identifier": "import-21-function1",
        },
      ],
    },
  ],
  "name": "this-project",
  "relationships": Array [
    Object {
      "from": "filename-7.js",
      "to": "external-7",
    },
    Object {
      "from": "filename-11.js",
      "to": "external-11",
    },
    Object {
      "from": "filename-17.js",
      "to": "external-17",
    },
    Object {
      "from": "filename-21.js",
      "to": "external-21",
    },
  ],
}
`;

exports[`fp-diagram should convert diagram to a diagram structure 2`] = `
"classDiagram
class \`filename-7.js\`{
  +create-7()
  -update-7()
  +delete-7()
}
class \`filename-11.js\`{
  +create-11()
  -update-11()
  +delete-11()
}
class \`filename-17.js\`{
  +create-17()
  -update-17()
  +delete-17()
}
class \`filename-21.js\`
class \`external-7\`{
  +import-7-function1()
}
class \`external-11\`{
  +import-11-function1()
}
class \`external-17\`{
  +import-17-function1()
}
class \`external-21\`{
  +import-21-function1()
}
\`filename-7.js\`-->\`external-7\`
\`filename-11.js\`-->\`external-11\`
\`filename-17.js\`-->\`external-17\`
\`filename-21.js\`-->\`external-21\`"
`;
